packages/mobile/src/firebase/firebase.ts

```diff
@@ -26,57 +46,68 @@ export const initializeAuth = async (app: Firebase, address: string) => {
   Logger.info(TAG, 'Firebase Auth initialized successfully')
 }
 
-export const initializeCloudMessaging = async (app: Firebase, address: string) => {
-  // TODO(cmcewen): remove once we move off thunk
-  const store = getReduxStore()
-  const language = store.getState().app.language
-  const dispatch = store.dispatch
-
+export function* initializeCloudMessaging(app: Firebase, address: string) {
   Logger.info(TAG, 'Initializing Firebase Cloud Messaging')
-  const enabled = await app.messaging().hasPermission()
+
+  const enabled = yield call([app.messaging(), 'hasPermission'])
   if (!enabled) {
     try {
-      await app.messaging().requestPermission()
+      yield call([app.messaging(), 'requestPermission'])
     } catch (error) {
       Logger.error(TAG, 'User has rejected messaging permissions', error)
       throw error
     }
   }
 
-  const fcmToken = await app.messaging().getToken()
+  const fcmToken = yield call([app.messaging(), 'getToken'])
   if (fcmToken) {
-    await registerTokenToDb(app, address, fcmToken)
+    yield call(registerTokenToDb, app, address, fcmToken)
     // First time setting the fcmToken also set the language selection
-    await setUserLanguage(address, language)
+    const language = yield select(currentLanguageSelector)
+    yield call(setUserLanguage, address, language)
   }
 
-  // Monitor for future token refreshes
   app.messaging().onTokenRefresh(async (token) => {
     Logger.info(TAG, 'Cloud Messaging token refreshed')
     await registerTokenToDb(app, address, token)
   })
 
   // Listen for notification messages while the app is open
-  app.notifications().onNotification((notification: Notification) => {
-    Logger.info(TAG, 'Notification received while open')
-    dispatch(handleNotification(notification, NotificationReceiveState.APP_ALREADY_OPEN))
-  })
-
-  app.notifications().onNotificationOpened((notification: NotificationOpen) => {
-    Logger.info(TAG, 'App opened via a notification')
-    dispatch(
-      handleNotification(notification.notification, NotificationReceiveState.APP_FOREGROUNDED)
+  const channelOnNotification: EventChannel<{
+    notification: Notification
+    stateType: NotificationReceiveState
+  }> = eventChannel((emitter) => {
+    app.notifications().onNotification(
+      (notification: Notification): any => {
+        Logger.info(TAG, 'Notification received while open')
+        emitter({ notification, stateType: NotificationReceiveState.APP_ALREADY_OPEN })
+      }
     )
+    return () => null // Return an unsubscribe method
+  })
+  spawn(watchFirebaseNotificationChannel, channelOnNotification)
```

Comment:
Please use fork here instead of spawn.
The difference is that spawn creates an error boundary but actually it would be good for notification related errors to all trace back to here IMO

Reply:
But if I fork, wont it be killed after I `initializeCloudMessaging` finishes?